coderd/keyrotate/rotate.go

```diff
@@ -0,0 +1,287 @@
+package keyrotate
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"encoding/hex"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/quartz"
+)
+
+const (
+	WorkspaceAppsTokenDuration = time.Minute
+	OIDCConvertTokenDuration   = time.Minute * 5
+	TailnetResumeTokenDuration = time.Hour * 24
+
+	// DefaultRotationInterval is the default interval at which keys are checked for rotation.
+	DefaultRotationInterval = time.Minute * 10
+	// DefaultKeyDuration is the default duration for which a key is valid. It applies to all features.
+	DefaultKeyDuration = time.Hour * 24 * 30
+)
+
+// Rotator is responsible for rotating keys in the database.
+type Rotator struct {
+	db          database.Store
+	logger      slog.Logger
+	clock       quartz.Clock
+	keyDuration time.Duration
+	// resultsCh is purely for testing.
+	resultsCh chan []database.CryptoKey
+
+	// The following fields are instantiated in Open.
+	ticker   *quartz.Ticker
+	features []database.CryptoKeyFeature
+}
+
+// Open instantiates a new Rotator. It ensures there's at least one
+// valid key per feature prior to returning. Close should be called
+// to ensure the ticker that is instantiated is not leaked.
+func Open(ctx context.Context, db database.Store, logger slog.Logger, clock quartz.Clock, rotateInterval time.Duration, keyDuration time.Duration, results chan []database.CryptoKey) (*Rotator, error) {
```

Comment:
it would be cleaner if the quartz clock were moved to an `opts ...Option` style argument.